File: apps/demo-ios/Sources/Models/AppSettings.swift
import Foundation
import SEACore

/// Persisted tester-facing configuration for the demo harness (UserDefaults-backed).
///
/// This is harness state only — it has no bearing on SEACore's own security
/// posture. `allowedDomains` here is the *host-supplied narrowing list* that
/// SEACore intersects with `SEAEnvironment.current.authDomains` (loaded from
/// this app's own bundled `SEASecurityConfig.plist` — api-contract-ios-v1.md
/// §3.3) — it can never widen what the core accepts.
final class AppSettings: ObservableObject {
    static let shared = AppSettings()

    private enum Keys {
        static let gatewayBaseURL = "sea.demo.gatewayBaseURL"
        static let allowedDomains = "sea.demo.allowedDomains"
        static let presentation = "sea.demo.presentation"
        static let timeoutMs = "sea.demo.timeoutMs"
        static let useMockGateway = "sea.demo.useMockGateway"
        static let mockRedirectURL = "sea.demo.mockRedirectURL"
    }

    private enum Defaults {
        // Our own GatewayClient <-> local backend traffic. Not passed to SEACore.
        static let gatewayBaseURL = "http://localhost:8080"
        // Narrowing list handed to SEACore; also reused by the fuzz screen so
        // both surfaces exercise the same effective allowlist.
        static let allowedDomains = "auth.bank.local,localhost,platform-keycloak.pres.proxym-it.net"
        static let presentation = "sheet"
        static let timeoutMs = 120_000
        static let useMockGateway = true
        // A plausible local-Keycloak authorize URL. Must be https + a host in
        // the allowlist above, or SEACore's validator will reject it by design
        // (api-contract-ios-v1.md §5, rule 1 and rule 5) — that's the point of
        // the mock: it lets you exercise the WebView surface, but it does not
        // and should not bypass the core's own validation.
        // A COMPLETE, working authorize URL for the local infra/ Keycloak:
        //   - host auth.bank.local (TLS-terminated; in this app's
        //     SEASecurityConfig.plist allowlist). This host is what backs the
        //     passkey RP ID / Associated Domain (§10), so the WebView MUST load
        //     from it — not localhost — for a WebAuthn ceremony to bind. It
        //     resolves to 127.0.0.1 via the Mac's /etc/hosts (the Simulator
        //     inherits it); see infra/README.md for the one-time sudo line.
        //   - realm bankerise-mobile, client sea-dev-public (the dev-only
        //     public client — see infra/README.md)
        //   - redirect_uri bkrmob://callback — the scheme SEACore actually
        //     captures (SEASecurityConfig.plist CallbackScheme). It must match
        //     the client's registered redirectUris in provision-realm.sh.
        //   - PKCE S256 params, which sea-dev-public enforces. Omitting them
        //     makes Keycloak 302 straight to bkrmob://callback?error=…
        //     which SEA then correctly *captures* as an error-shaped callback
        //     (§6.3) instead of ever showing a login form. The challenge below
        //     is the RFC 7636 worked example (verifier
        //     dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk).
        static let mockRedirectURL =
            "https://auth.bank.local/realms/bankerise-mobile/protocol/openid-connect/auth" +
            "?client_id=sea-dev-public&redirect_uri=bkrmob%3A%2F%2Fcallback" +
            "&response_type=code&scope=openid&state=devstate123" +
            "&code_challenge=E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM" +
            "&code_challenge_method=S256"
    }

    private let defaults = UserDefaults.standard

    @Published var gatewayBaseURL: String {
        didSet { defaults.set(gatewayBaseURL, forKey: Keys.gatewayBaseURL) }
    }
    @Published var allowedDomains: String {
        didSet { defaults.set(allowedDomains, forKey: Keys.allowedDomains) }
    }
    @Published var presentation: SEAPresentation {
        didSet {
            // Avoid relying on Equatable conformance for SEAPresentation (not
            // documented in the contract) — switch instead of `==`.
            let raw: String
            switch presentation {
            case .sheet: raw = "sheet"
            case .fullscreen: raw = "fullscreen"
            }
            defaults.set(raw, forKey: Keys.presentation)
        }
    }
    @Published var timeoutMs: Int {
        didSet { defaults.set(timeoutMs, forKey: Keys.timeoutMs) }
    }
    @Published var useMockGateway: Bool {
        didSet { defaults.set(useMockGateway, forKey: Keys.useMockGateway) }
    }
    @Published var mockRedirectURL: String {
        didSet { defaults.set(mockRedirectURL, forKey: Keys.mockRedirectURL) }
    }

    private init() {
        gatewayBaseURL = defaults.string(forKey: Keys.gatewayBaseURL) ?? Defaults.gatewayBaseURL
        allowedDomains = defaults.string(forKey: Keys.allowedDomains) ?? Defaults.allowedDomains
        let presentationRaw = defaults.string(forKey: Keys.presentation) ?? Defaults.presentation
        presentation = presentationRaw == "fullscreen" ? .fullscreen : .sheet
        let storedTimeout = defaults.object(forKey: Keys.timeoutMs) as? Int
        timeoutMs = storedTimeout ?? Defaults.timeoutMs
        if defaults.object(forKey: Keys.useMockGateway) != nil {
            useMockGateway = defaults.bool(forKey: Keys.useMockGateway)
        } else {
            useMockGateway = Defaults.useMockGateway
        }
        mockRedirectURL = defaults.string(forKey: Keys.mockRedirectURL) ?? Defaults.mockRedirectURL
    }

    /// Comma list -> trimmed, non-empty array, in the shape SEAConfig.allowedDomains
    /// and the fuzz screen both expect.
    var allowedDomainsArray: [String] {
        allowedDomains
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}

File: apps/demo-android/app/src/main/java/com/bankerise/sea/demo/model/AppSettings.kt
package com.bankerise.sea.demo.model

import android.content.Context
import android.content.SharedPreferences
import com.bankerise.sea.core.SEACapturePolicy
import com.bankerise.sea.core.SEAConfig
import com.bankerise.sea.core.SEAPresentation
import android.net.Uri

/**
 * SharedPreferences-backed tester-facing configuration for the demo harness.
 *
 * This is harness state only — it has no bearing on SEACore's own security
 * posture. `allowedDomains` here is the *host-supplied narrowing list* that
 * SEACore intersects with its compiled [com.bankerise.sea.core.SEAEnvironment.authDomains]
 * — it can never widen what the core accepts.
 */
class AppSettings private constructor(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("sea_demo", Context.MODE_PRIVATE)

    var gatewayBaseUrl: String
        get() = prefs.getString(KEY_GATEWAY_BASE_URL, DEFAULTS.gatewayBaseUrl) ?: DEFAULTS.gatewayBaseUrl
        set(value) = prefs.edit().putString(KEY_GATEWAY_BASE_URL, value).apply()

    var callbackScheme: String
        get() = prefs.getString(KEY_CALLBACK_SCHEME, DEFAULTS.callbackScheme) ?: DEFAULTS.callbackScheme
        set(value) = prefs.edit().putString(KEY_CALLBACK_SCHEME, value).apply()

    var allowedDomains: String
        get() = prefs.getString(KEY_ALLOWED_DOMAINS, DEFAULTS.allowedDomains) ?: DEFAULTS.allowedDomains
        set(value) = prefs.edit().putString(KEY_ALLOWED_DOMAINS, value).apply()

    var presentation: String
        get() = prefs.getString(KEY_PRESENTATION, DEFAULTS.presentation) ?: DEFAULTS.presentation
        set(value) = prefs.edit().putString(KEY_PRESENTATION, value).apply()

    var timeoutMs: Long
        get() = prefs.getLong(KEY_TIMEOUT_MS, DEFAULTS.timeoutMs)
        set(value) = prefs.edit().putLong(KEY_TIMEOUT_MS, value).apply()

    var useMockGateway: Boolean
        get() = prefs.getBoolean(KEY_USE_MOCK_GATEWAY, DEFAULTS.useMockGateway)
        set(value) = prefs.edit().putBoolean(KEY_USE_MOCK_GATEWAY, value).apply()

    var mockRedirectUrl: String
        get() = prefs.getString(KEY_MOCK_REDIRECT_URL, DEFAULTS.mockRedirectUrl) ?: DEFAULTS.mockRedirectUrl
        set(value) = prefs.edit().putString(KEY_MOCK_REDIRECT_URL, value).apply()

    /**
     * Comma list → trimmed, non-empty array, in the shape
     * [SEAConfig.allowedDomains] expects.
     */
    val allowedDomainsArray: List<String>
        get() = allowedDomains
            .split(",")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

    fun toSEAConfig(authorizeUrl: Uri): SEAConfig {
        val pres = try {
            SEAPresentation.valueOf(presentation.uppercase())
        } catch (_: Exception) { SEAPresentation.SHEET }

        return SEAConfig(
            authorizeUrl = authorizeUrl,
            callbackScheme = callbackScheme,
            allowedDomains = allowedDomainsArray,
            presentation = pres,
            timeoutMs = timeoutMs
        )
    }

    companion object {
        @Volatile
        private var instance: AppSettings? = null

        fun getInstance(context: Context): AppSettings {
            return instance ?: synchronized(this) {
                instance ?: AppSettings(context).also { instance = it }
            }
        }

        private object KEYS {
            const val GATEWAY_BASE_URL = "sea.demo.gatewayBaseURL"
            const val CALLBACK_SCHEME = "sea.demo.callbackScheme"
            const val ALLOWED_DOMAINS = "sea.demo.allowedDomains"
            const val PRESENTATION = "sea.demo.presentation"
            const val TIMEOUT_MS = "sea.demo.timeoutMs"
            const val USE_MOCK_GATEWAY = "sea.demo.useMockGateway"
            const val MOCK_REDIRECT_URL = "sea.demo.mockRedirectURL"
        }
        private const val KEY_GATEWAY_BASE_URL = KEYS.GATEWAY_BASE_URL
        private const val KEY_CALLBACK_SCHEME = KEYS.CALLBACK_SCHEME
        private const val KEY_ALLOWED_DOMAINS = KEYS.ALLOWED_DOMAINS
        private const val KEY_PRESENTATION = KEYS.PRESENTATION
        private const val KEY_TIMEOUT_MS = KEYS.TIMEOUT_MS
        private const val KEY_USE_MOCK_GATEWAY = KEYS.USE_MOCK_GATEWAY
        private const val KEY_MOCK_REDIRECT_URL = KEYS.MOCK_REDIRECT_URL

        private object DEFAULTS {
            const val gatewayBaseUrl = "https://auth-retail.demo.proxym-it.net"
            const val callbackScheme = "bankerise-auth"
            const val allowedDomains = "auth.bank.local,localhost,10.0.2.2,auth-retail.demo.proxym-it.net"
            const val presentation = "SHEET"
            const val timeoutMs = 120_000L
            const val useMockGateway = true
            const val mockRedirectUrl =
                "https://auth-retail.demo.proxym-it.net/realms/uib-qa-front/protocol/openid-connect/auth?response_type=code&client_id=Customer&scope=openid%20profile%20offline_access&state=GJlqE4BlC6kXKhNulrnrDmekbB2JC9JqL5N_sjL8K_s%3D&redirect_uri=http://uib-ebanking.test.proxym-it.tn/login/oauth2/code/Customer&nonce=Yotlp0gGrSVk5VHXJVWYsRtuiKl8BiOvSdbSLXgx0X8&code_challenge=Y_urVAne_932wZU8z6XrZrCg5l9Q4eOfron0Cma67bU&code_challenge_method=S256&ui_locales=fr&mode=light"
        }
    }
}
